File: csharp/solid/OpenClosed.cs
namespace Solid
{
    ///
    /// <summary>
    /// Class <c>OpenClosed</c> represents Open-Closed principle usecase.
    /// 1. classes, modules, functions should be open for extension, but closed for modification
    /// 2. such an entity can allow its behaviour to be extended without modifying its source code
    /// </summary>
    ///
    class OpenClosed
    {
        public static void Run()
        {
            Console.WriteLine("\nOpen Closed\n");

            // create products
            var candy = new Product("Candy", Color.Red, Size.Small);
            var dress = new Product("Dress", Color.Red, Size.Medium);
            var book = new Product("Book", Color.Green, Size.Medium);

            // create array of products
            Product[] products = { candy, dress, book };

            var pf = new ProductFiler();
            Console.WriteLine("Filter Products (Violating Open-Closed Principle):");
            foreach (var p in pf.FilterByColor(products, Color.Red))
            {
                Console.WriteLine($"Name: {p.Name}, Color: {p.Color}, Size: {p.Size}");
            }
            foreach (var p in pf.FilterBySize(products, Size.Medium))
            {
                Console.WriteLine($"Name: {p.Name}, Color: {p.Color}, Size: {p.Size}");
            }

            var pfe = new ProductFilterEnhanced();
            Console.WriteLine("Filter Products (Utilizing Open-Closed Principle):");
            foreach (var p in pfe.Filter(products, new ColorSpecification(Color.Red)))
            {
                Console.WriteLine($"Name: {p.Name}, Color: {p.Color}, Size: {p.Size}");
            }
            foreach (var p in pfe.Filter(products, new SizeSpecification(Size.Medium)))
            {
                Console.WriteLine($"Name: {p.Name}, Color: {p.Color}, Size: {p.Size}");
            }
            foreach (var p in pfe.Filter(
                products,
                new AndSpecification<Product>(
                    new ColorSpecification(Color.Red),
                    new SizeSpecification(Size.Small))
                ))
            {
                Console.WriteLine($"Name: {p.Name}, Color: {p.Color}, Size: {p.Size}");
            }
        }
    }

    /// <summary>Class <c>Color</c> represents available colors.</summary>
    public enum Color
    {
        Red, Green, Blue
    }

    /// <summary>Class <c>Size</c>  represents available sizes</summary>
    public enum Size
    {
        Small, Medium, Large
    }

    /// <summary>Class <c>Product</c> represents product entity.</summary>
    public class Product
    {
        public string Name;
        public Color Color;
        public Size Size;

        public Product(string name, Color color, Size size)
        {
            Name = name ?? throw new ArgumentNullException(paramName: nameof(name));
            Color = color;
            Size = size;
        }
    }

    /// <summary>Class <c>ProductFilter</c> concerned about filtering products by certain criteria.</summary>
    public class ProductFiler
    {
        // Coming back to modify filter class every time business rules change would violate Open-Closed principle.

        /// <summary>Method <c>FilterBySize</c> filter products by size.</summary>
        public IEnumerable<Product> FilterBySize(IEnumerable<Product> products, Size size)
        {
            foreach (var p in products)
            {
                if (p.Size == size)
                {
                    yield return p;
                }
            }
        }

        /// <summary>Method <c>FilterByColor</c> filter products by color.</summary>
        public IEnumerable<Product> FilterByColor(IEnumerable<Product> products, Color color)
        {
            foreach (var p in products)
            {
                if (p.Color == color)
                {
                    yield return p;
                }
            }
        }

        /// <summary>Method <c>FilterByColor</c> filter products by size and color...</summary>
    }

    /// <summary>Interface <c>ISpecification</c> defines specification interface.</summary>
    public interface ISpecification<T>
    {
        bool IsSatisfied(T t);
    }

    /// <summary>Interface <c>IFilter</c> defines filter interface.</summary>
    public interface IFilter<T>
    {
        IEnumerable<T> Filter(IEnumerable<T> items, ISpecification<T> spec);
    }

    /// <summary>Class <c>ColorSpecification</c> implements specification interface.</summary>
    public class ColorSpecification : ISpecification<Product>
    {
        private Color color;
        public ColorSpecification(Color color)
        {
            this.color = color;
        }
        public bool IsSatisfied(Product p)
        {
            return color == p.Color;
        }
    }

    /// <summary>Class <c>SizeSpecification</c> implements specification interface.</summary>
    public class SizeSpecification : ISpecification<Product>
    {
        private Size size;
        public SizeSpecification(Size size)
        {
            this.size = size;
        }
        public bool IsSatisfied(Product p)
        {
            return size == p.Size;
        }
    }

    /// <summary>Class <c>AndSpecification</c> implements specification interface as well implements combinator of specifications.</summary>
    public class AndSpecification<T> : ISpecification<T>
    {
        private ISpecification<T> first, second;

        public AndSpecification(ISpecification<T> first, ISpecification<T> second)
        {
            this.first = first ?? throw new ArgumentNullException(paramName: nameof(first));
            this.second = second ?? throw new ArgumentNullException(paramName: nameof(second));
        }

        public bool IsSatisfied(T t)
        {
            return first!.IsSatisfied(t) && second!.IsSatisfied(t);
        }
    }

    /// <summary>Class <c>ProductFilterEnhanced</c> concerned about filtering products by certain criteria utilizing open-closed principle.</summary>
    public class ProductFilterEnhanced : IFilter<Product>
    {
        public IEnumerable<Product> Filter(IEnumerable<Product> items, ISpecification<Product> spec)
        {
            foreach (var i in items)
            {
                if (spec.IsSatisfied(i))
                {
                    yield return i;
                }
            }
        }
    }
}

File: golang/solid/open_closed.go
package solid

import "fmt"

// O - Open-Closed (open for extension, closed for modification)

// 1. should be open for extension, but closed for modification
// 2. such an entity can allow its behaviour to be extended without modifying its source code

func OpenClosed() {
	fmt.Println("Open-Closed")

	// init range of products
	pp := []Product{
		{
			Name:  "Little Black Dress",
			Color: ProductColorBlack,
			Size:  ProductSizeSmall,
		},
		{
			Name:  "Sword of Isildur",
			Color: ProductColorMetallic,
			Size:  ProductSizeLarge,
		},
		{
			Name:  "BMW M3 GTR",
			Color: ProductColorBlack,
			Size:  ProductSizeLarge,
		},
	}
	fmt.Printf("%+v\n\n", pp)

	// Not Good

	// init product filter
	f := NewProductFilter()

	largeProducts := f.FilterBySize(pp, ProductSizeLarge)
	blackProducts := f.FilterByColor(pp, ProductColorBlack)

	fmt.Printf("[B] Large Products: %+v\n", largeProducts)
	fmt.Printf("[B] Black Products: %+v\n\n", blackProducts)

	// Good

	fe := NewProductFilterEnhanced()

	largeProds := fe.FilterProducts(pp, NewProductSizeSpecification(ProductSizeLarge))
	blackProds := fe.FilterProducts(pp, NewProductColorSpecification(ProductColorBlack))
	blackAndLargeProds := fe.FilterProducts(pp,
		NewProductAndSpecification(
			NewProductSizeSpecification(ProductSizeLarge),
			NewProductColorSpecification(ProductColorBlack),
		),
	)

	fmt.Printf("[G] Large Products: %+v\n", largeProds)
	fmt.Printf("[G] Black Products: %+v\n", blackProds)
	fmt.Printf("[G] Large Black Products: %+v\n\n", blackAndLargeProds)

	fee := ProductFilterEnhanced{}
	largeProdse := fee.FilterProducts(pp, &ProductSizeSpecification{ProductSizeLarge})
	blackProdse := fee.FilterProducts(pp, &ProductColorSpecification{ProductColorBlack})
	blackAndLargeProdse := fee.FilterProducts(pp,
		&ProductAndSpecification{
			[]ProductSpecification{
				&ProductSizeSpecification{ProductSizeLarge},
				&ProductColorSpecification{ProductColorBlack},
			},
		},
	)

	fmt.Printf("[GE] Large Products: %+v\n", largeProdse)
	fmt.Printf("[GE] Black Products: %+v\n", blackProdse)
	fmt.Printf("[GE] Large Black Products: %+v\n", blackAndLargeProdse)
}

// Product - represents product.
type Product struct {
	Name  string
	Color ProductColor
	Size  ProductSize
}

// ProductColor - represents Product colors.
type ProductColor string

const (
	ProductColorBlack    ProductColor = "black"
	ProductColorBlue     ProductColor = "blue"
	ProductColorMetallic ProductColor = "metallic"
)

// ProductSize - represents Product sizes.
type ProductSize string

const (
	ProductSizeSmall  ProductSize = "S"
	ProductSizeMedium ProductSize = "M"
	ProductSizeLarge  ProductSize = "L"
)

// ProductFilter - represents ProductFilter.
type ProductFilter struct{}

// NewProductFilter - creates new instance of ProductFilter.
func NewProductFilter() *ProductFilter {
	return &ProductFilter{}
}

// FilterBySize - filters Product's by size.
func (pf *ProductFilter) FilterBySize(pp []Product, s ProductSize) []Product {
	result := make([]Product, 0)
	for _, p := range pp {
		if p.Size == s {
			result = append(result, p)
		}
	}
	return result
}

// FilterByColor - filters Product's by color.
func (pf *ProductFilter) FilterByColor(pp []Product, c ProductColor) []Product {
	result := make([]Product, 0)
	for _, p := range pp {
		if p.Color == c {
			result = append(result, p)
		}
	}
	return result
}

// FilterByColorAndSize
// FilterBy...
// FilterBy...

// Specification Pattern

// ProductSpecification - represents Product specification interface.
type ProductSpecification interface {
	ProductSatisfied(p Product) bool
}

// ProductsFilter - represents Product's filter interface.
type ProductsFilter interface {
	FilterProducts(pp []Product, s ProductSpecification) []Product
}

// Color Specification

// ProductColorSpecification - implements ProductSpecification interface.
type ProductColorSpecification struct {
	Color ProductColor
}

// NewProductColorSpecification - creates new instance of ProductColorSpecification.
func NewProductColorSpecification(color ProductColor) *ProductColorSpecification {
	return &ProductColorSpecification{color}
}

// ProductSatisfied - ensures that Product specification is satisfied.
func (pcs *ProductColorSpecification) ProductSatisfied(p Product) bool {
	return p.Color == pcs.Color
}

// Size Specification

// ProductSizeSpecification - implements ProductSpecification interface.
type ProductSizeSpecification struct {
	Size ProductSize
}

// NewProductSizeSpecification - creates new instance of ProductSizeSpecification.
func NewProductSizeSpecification(size ProductSize) *ProductSizeSpecification {
	return &ProductSizeSpecification{size}
}

// ProductSatisfied - ensures that Product specification is satisfied.
func (pss *ProductSizeSpecification) ProductSatisfied(p Product) bool {
	return p.Size == pss.Size
}

// And Specification

// ProductAndSpecification - implements ProductSpecification interface.
// combinator of ProductSpecifications.
type ProductAndSpecification struct {
	Specifications []ProductSpecification
}

// NewProductAndSpecification - creates new instance of ProductAndSpecification.
func NewProductAndSpecification(specs ...ProductSpecification) *ProductAndSpecification {
	return &ProductAndSpecification{specs}
}

// ProductSatisfied - ensures that Product specification is satisfied.
func (pas *ProductAndSpecification) ProductSatisfied(p Product) bool {
	for i := range pas.Specifications {
		if !pas.Specifications[i].ProductSatisfied(p) {
			return false
		}
	}
	return true
}

// Product Filter

// ProductFilterEnhanced - implements ProductsFilter interface.
type ProductFilterEnhanced struct{}

// NewProductFilterEnhanced - creates new instance of ProductFilterEnhanced.
func NewProductFilterEnhanced() *ProductFilterEnhanced {
	return &ProductFilterEnhanced{}
}

// FilterProducts - filters range of products based on ProductSpecification.
func (pfe *ProductFilterEnhanced) FilterProducts(pp []Product, s ProductSpecification) []Product {
	result := make([]Product, 0)
	for _, p := range pp {
		if s.ProductSatisfied(p) {
			result = append(result, p)
		}
	}
	return result
}
